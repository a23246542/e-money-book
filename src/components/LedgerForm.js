import React from 'react'
import PropTypes from 'prop-types'

const LedgerForm = props => {
  return (
    <div>
      表單
    </div>
  )
}

LedgerForm.propTypes = {

}

export default LedgerForm
