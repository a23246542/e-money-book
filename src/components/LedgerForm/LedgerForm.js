import React, { useState, useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import style from './style.module.scss';

const LedgerFormComponent = ({
  ledgerItem = {},
  onFormSubmit,
  onCancelSubmit,
  children,
}) => {
  const [title, setTitle] = useState('');
  const [amount, setAmount] = useState('');
  const [date, setDate] = useState('');
  const [formValidate, setFormValidate] = useState({
    validatePass: true,
    alertMessage: '',
  });

  useEffect(() => {
    if (!ledgerItem.id) {
      return;
    }
    setTitle(ledgerItem.title);
    setAmount(ledgerItem.amount.toString());
    setDate(ledgerItem.date);
  }, [ledgerItem.id, ledgerItem.title, ledgerItem.amount, ledgerItem.date]);

  const isValidDate = (inputDate) => {
    const nowTimeStamp = new Date().getTime();
    const selectTimeStamp = new Date(inputDate).setHours(0, 0, 0, 0);
    return selectTimeStamp <= nowTimeStamp;
  };

  const submitForm = (e) => {
    e.preventDefault();
    const isEditMode = !!ledgerItem.id;
    const amountNumber = Number(amount);
    if (title && amount && date) {
      if (amountNumber < 0) {
        setFormValidate({
          validatePass: false,
          alertMessage: '數字不能為負',
        });
      } else if (!isValidDate(date)) {
        setFormValidate({
          validatePass: false,
          alertMessage: '不能選擇未來的日期',
        });
      } else {
        setFormValidate({
          validatePass: true,
          alertMessage: '',
        });
        if (isEditMode) {
          onFormSubmit(
            {
              ...ledgerItem,
              title,
              amount: amountNumber,
              date,
            },
            isEditMode
          );
          setTitle('');
          setAmount('');
          setDate('');
        } else {
          onFormSubmit(
            {
              title,
              amount: amountNumber,
              date,
            },
            isEditMode
          );
          setTitle('');
          setAmount('');
          setDate('');
        }
      }
    } else {
      setFormValidate({
        validatePass: false,
        alertMessage: '表格不能為空',
      });
    }
  };

  const cancelSubmit = (e) => {
    e.preventDefault();
    onCancelSubmit();
  };

  return (
    <div className="container" data-testid="LedgerForm">
      <form className="pt-5 px-2">
        <div className="form-group">
          <div className="row">
            <div className="col-2 col-form-label">
              <label htmlFor="inputTitle" className="text-nowrap">
                標題*:{' '}
              </label>
            </div>
            <div className="col-10">
              <input
                type="text"
                className="form-control"
                id="inputTitle"
                data-testid="inputTitle"
                aria-describedby="emailHelp"
                value={title || ''}
                onChange={(e) => {
                  setTitle(e.target.value.trim());
                }}
              />
            </div>
          </div>
        </div>
        <div className="form-group">
          <div className="row">
            <div className="col-2 col-form-label">
              <label htmlFor="inputAmount" className="text-nowrap">
                金額*:
              </label>
            </div>
            <div className="col-10">
              <div className={`${style.inputAmount} input-group`}>
                <div className="input-group-prepend">
                  <span className="input-group-text">$</span>
                </div>
                <input
                  type="number"
                  className="form-control"
                  id="inputAmount"
                  data-testid="inputAmount"
                  value={amount || ''}
                  onChange={(e) => {
                    setAmount(e.target.value.trim().toString());
                  }} //@@
                />
              </div>
            </div>
          </div>
        </div>
        <div className="form-group">
          <div className="row">
            <div className="col-2 col-form-label">
              <label htmlFor="inputDate" className="text-nowrap">
                日期*:
              </label>
            </div>
            <div className="col-10">
              <input
                type="date"
                className="form-control"
                id="inputDate"
                data-testid="inputDate"
                value={date || ''}
                onChange={(e) => {
                  setDate(e.target.value.trim());
                }}
                // ref={(input) => {setDate(input.current.value)}}
              />
            </div>
          </div>
        </div>
        {!formValidate.validatePass && (
          <div className="alert alert-warning" role="alert">
            {formValidate.alertMessage}
          </div>
        )}
        {children}
        <button
          type="submit"
          id="submit"
          data-testid="submit"
          className="btn btn-primary mx-3"
          onClick={submitForm}
        >
          提交
        </button>
        <button
          type="button"
          id="cancel"
          data-testid="cancel"
          className="btn btn-primary mx-3"
          onClick={(e) => {
            cancelSubmit(e);
          }}
        >
          取消
        </button>
      </form>
    </div>
  );
};

LedgerFormComponent.propTypes = {
  ledgerItem: PropTypes.object,
  onFormSubmit: PropTypes.func.isRequired,
  onCancelSubmit: PropTypes.func.isRequired,
};

export const LedgerForm = memo(LedgerFormComponent);
