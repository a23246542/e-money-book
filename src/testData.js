import { TYPE_OUTCOME, TYPE_INCOME } from './constants';

export const testTabs = [
  {
    text: '支出',
    value: TYPE_OUTCOME,
    iconName: 'IosCard',
  },
  {
    text: '收入',
    value: TYPE_INCOME,
    iconName: 'IosCash',
  },
  // {
  //   text:'其他',
  //   value:'other',
  //   iconName:'IosCash'
  // },
];

export const testCategories = [
  {
    name: '旅行',
    iconName: 'IosPlane',
    id: '1',
    type: 'outcome',
  },
  {
    name: '餐饮',
    iconName: 'IosRestaurant',
    id: '2',
    type: 'outcome',
  },
  {
    name: '购物',
    iconName: 'IosBasket',
    id: '3',
    type: 'outcome',
  },
  {
    name: '数码',
    iconName: 'IosPhonePortrait',
    id: '4',
    type: 'outcome',
  },
  {
    name: '工资',
    iconName: 'IosCard',
    id: '10',
    type: 'income',
  },
  {
    name: '兼职',
    iconName: 'IosCash',
    id: '11',
    type: 'income',
  },
  {
    name: '理财',
    iconName: 'LogoYen',
    id: '12',
    type: 'income',
  },
];

export const testItems = [
  {
    title: '工资',
    amount: 2000,
    date: '2020-11-18',
    monthCategory: '2020-11',
    id: '_bd16bjeen',
    cid: '10',
    timestamp: 1534550400000,
  },
  {
    title: '和哥们一起喝酒',
    amount: 300,
    date: '2020-12-20',
    monthCategory: '2020-12',
    id: '_jjfice21k',
    cid: '3',
    timestamp: 1534723200000,
  },
  {
    title: '抽奖',
    amount: 1000,
    date: '2020-12-11',
    monthCategory: '2020-12',
    id: '_1fg1wme63',
    cid: '11',
    timestamp: 1533945600000,
  },
  {
    title: '理财收入',
    amount: 300,
    date: '2020-11-15',
    monthCategory: '2020-11',
    id: '_qryggm5y8',
    cid: '12',
    timestamp: 1534291200000,
  },
  {
    title: '请别人吃饭',
    amount: 300,
    date: '2020-12-15',
    monthCategory: '2020-12',
    id: '_qryggm511',
    cid: '3',
    timestamp: 1534291200000,
  },
];
export const testItemsCopy = [
  {
    title: '工资',
    amount: 2000,
    date: '2020-11-18',
    monthCategory: '2020-11',
    id: '_bd16bjeen',
    cid: '10',
    timestamp: 1534550400000,
  },
  {
    title: '喝酒',
    amount: 300,
    date: '2020-12-20',
    monthCategory: '2020-12',
    id: '_jjfice21k',
    cid: '3',
    timestamp: 1534723200000,
  },
  {
    title: '抽奖',
    amount: 1000,
    date: '2020-12-11',
    monthCategory: '2020-12',
    id: '_1fg1wme63',
    cid: '11',
    timestamp: 1533945600000,
  },
  {
    title: '理财收入',
    amount: 300,
    date: '2020-11-15',
    monthCategory: '2020-11',
    id: '_qryggm5y8',
    cid: '12',
    timestamp: 1534291200000,
  },
  {
    title: '请别人吃饭',
    amount: 300,
    date: '2020-12-15',
    monthCategory: '2020-12',
    id: '_qryggm511',
    cid: '3',
    timestamp: 1534291200000,
  },
];
