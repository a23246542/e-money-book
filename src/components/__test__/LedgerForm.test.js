import React from 'react';
import { shallow } from 'enzyme';
import { testItems } from '../../testData';
import LedgerForm from '../LedgerForm';

// 需求分析
//分創建跟編輯
//基本驗證內容
//提交跟取消會回調

//先props
const props = {
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn()
}

const props_with_item = {

}

let wrapper,  wrapper2, formInstance;


describe('test LedgerForm component', () => {
  beforeEach(()=>{
    wrapper = shallow(<LedgerForm {...props}/>);
    wrapper2 = shallow(<LedgerForm {...props_with_item}/>)
    // formInstance =
  })
  it('should render the component to match the snapshot',()=>{//%%
    expect(wrapper).toMatchSnapshot();
    expect(wrapper2).toMatchSnapshot();
  })
  //創建
  describe('test LedgerForm with no data ', () => {
    //-默認樣式
    //要看到三個input 一個form
    it('render should see three input and one form element',()=>{})
    // 要是空的
    it('render LedgerForm with no data should see three input and no value',()=>{})
    //-交互
    // 空值提交會出現報錯提示
    it('submit form with empty input should show alert message',()=>{})
    // 價格負的提交會錯
    it('submit form with invalid price should show alert message',()=>{})
    // 日期不符格式提交會錯
    it('submit form with invalid date should show alert message',()=>{})
    // 填寫正確提交會觸發回呼帶參數
    // it('submit form with valid data should invoke right function with right item',()=>{})
    it('submit form with valid data should trigger callback with right object',()=>{})
    // 點取消會觸發回調%%
    it('click the cancel button should call the right callback',()=>{});
  })

  //編輯
  describe('test LedgerForm with item data', () => {
    //默認
    //要看到三個input有資料
    // it('render the three input with value',()=>{})
    it('render LegerFrom with item should render correct data to input',()=>{})
    //交互
    // 都改ok後，會把對的資料物件呼叫出去
    // it('submit with valid data should call the right callback with right object',()=>{})
    it('submit with change value should trigger callback with right object',()=>{})
  })



})
