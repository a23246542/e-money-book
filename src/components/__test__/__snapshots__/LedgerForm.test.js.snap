// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test LedgerForm component should render the component to match the snapshot 1`] = `ShallowWrapper {}`;

exports[`test LedgerForm component should render the component to match the snapshot 2`] = `ShallowWrapper {}`;
